util: add vaultDestroyKeys option to destroy Vault kv-v2 secrets

nixpanic opened the pull request:
Hashicorp Vault does not completely remove the secrets in a kv-v2
backend when the keys are deleted. The metadata of the keys will be
kept, and it is possible to recover the contents of the keys afterwards.

With the new `vaultDestroyKeys` configuration parameter, this behaviour
can now be selected. By default the parameter will be set to `true`,
indicating that the keys and contents should completely be destroyed.
Setting it to any other value will make it possible to recover the
deleted keys.

See-also: https://github.com/libopenstorage/secrets/pull/55

---

<details>
<summary>Show available bot commands</summary>

These commands are normally not required, but in case of issues, leave any of
the following bot commands in an otherwise empty comment in this PR:

- `/retest ci/centos/<job-name>`: retest the `<job-name>` after unrelated
  failure (please report the failure too!)
- `/retest all`: run this in case the CentOS CI failed to start/report any test
  progress or results

</details>

nixpanic committed:
rebase: update libopenstorage/secrets

libopenstorage has added a new feature that makes it possible to destroy
the contents of a key/value in the Hashicorp Vault kv-v2 secrets backend.

See-also: https://github.com/libopenstorage/secrets/pull/55
Signed-off-by: Niels de Vos <ndevos@redhat.com>

nixpanic committed:
cleanup: use vaultDefaultCAVerify to set default value

Golang-ci complains about the following:

    internal/util/vault_tokens.go:99:20: string `true` has 4 occurrences, but such constant `vaultDefaultDestroyKeys` already exists (goconst)
    	v.VaultCAVerify = "true"
    	                  ^

This occurence of "true" can be replaced by vaultDefaultCAVerify so
address the warning.

Signed-off-by: Niels de Vos <ndevos@redhat.com>

nixpanic committed:
util: add vaultDestroyKeys option to destroy Vault kv-v2 secrets

Hashicorp Vault does not completely remove the secrets in a kv-v2
backend when the keys are deleted. The metadata of the keys will be
kept, and it is possible to recover the contents of the keys afterwards.

With the new `vaultDestroyKeys` configuration parameter, this behaviour
can now be selected. By default the parameter will be set to `true`,
indicating that the keys and contents should completely be destroyed.
Setting it to any other value will make it possible to recover the
deleted keys.

Signed-off-by: Niels de Vos <ndevos@redhat.com>

Madhu-1 commented:
@nixpanic CI failure.

Madhu-1 commented:
/test ci/centos/mini-e2e-helm/k8s-1.19

Madhu-1 commented:
> /test ci/centos/mini-e2e-helm/k8s-1.19

`X Exiting due to GUEST_PROVISION: Failed to start host: creating host: create: Error creating machine: Error in driver during machine creation: IP not available after waiting: machine minikube didn't return IP after 1 minute`

Madhu-1 reviewed (commented):
LGTM. @nixpanic can we update E2E to verify keys are deleted completely?

nixpanic committed:
e2e: add verifyKeyDestroyed() for validating `vaultDestroyKeys`

The kmsConfig type in the e2e suite has been enhanced with two functions
that make it possible to validate the destruction of deleted keys.

Signed-off-by: Niels de Vos <ndevos@redhat.com>

nixpanic commented:
> LGTM. @nixpanic can we update E2E to verify keys are deleted completely?

Sure, we default to kv-v2 in our testing, and that always supports the metadata for keys. A new commit has been added that enables checking for the metadata after deletion (should have been removed as well).

nixpanic commented:
@mergifyio rebase

#2349 has been merged, pulling the CentOS images should work again.

nixpanic commented:
/retest ci/centos/mini-e2e-helm/k8s-1.19

Madhu-1 commented on internal/util/vault.go:
```diff
@@ -268,6 +292,25 @@ func (vc *vaultConnection) Destroy() {
+// RemoveDEK() calls.
+func (vc *vaultConnection) getDeleteKeyContext() map[string]string {
+	keyContext := map[string]string{}
+	for k, v := range vc.keyContext {
+		keyContext[k] = v
+	}
```
@nixpanic do we need this new map generation or add key to existing map?

nixpanic replied:
Adding it to the existing map will cause failures during non-delete operations. So, creating a copy of the map and adding it there.

Madhu-1 commented:
/test ci/centos/mini-e2e-helm/k8s-1.20

Madhu-1 commented:
> /test ci/centos/mini-e2e-helm/k8s-1.20

`    Aug  3 14:41:07.037: timed out waiting for the rbd-nbd process: rbd-nbd process is not running yet: command terminated with exit code 1`

Madhu-1 commented:
@pkalever Looks like some nbd related failed https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/rest/organizations/jenkins/pipelines/mini-e2e-helm_k8s-1.20/runs/1891/nodes/110/steps/113/log/?start=0 PTAL

nixpanic commented:
Rebasing with resolving merge conflicts was needed. Please review again, @Madhu-1 @humblec

Rakshith-R commented on go.mod:
```diff
@@ -25,7 +25,7 @@ require (
 	k8s.io/api v0.21.2
 	k8s.io/apimachinery v0.21.2
-	k8s.io/client-go v0.21.2
+	k8s.io/client-go v12.0.0+incompatible
```
I am not sure, but should we be concerned about this ? 
@Madhu-1 @nixpanic @humblec

nixpanic replied:
I do not think so, but we can add the v0.21.2 version in the `replace` section if needed.

Rakshith-R replied:
`libopenstorage/secrets` has a `replace` internally too, let's add `k8s.io/client-go => k8s.io/client-go v0.21.2` here too ?
https://github.com/libopenstorage/secrets/blob/dde442ea20ec9d59c71cea5ee0f21eeffd17ed19/go.mod#L38

nixpanic replied:
Just wanted to add this, but it already is there :rofl: See below on line 48.

Rakshith-R replied:
Copy that.😂🤦‍♂️

nixpanic commented:
@mergifyio rebase
